sycl/doc/extensions/experimental/sycl_ext_oneapi_spec_constant_length_alloca.asciidoc

```diff
@@ -0,0 +1,266 @@
+submitted to the queue.
+
+== Overview
+
+{cpp} arrays and `std::array` can be used in SYCL code to represent fixed-size
+sequences of objects. However, these constructs have a significant restriction:
+the number of elements must be known at compile time. In host-code context,
+users can make use of dynamic memory allocations, e.g., `std::vector`, but this
+is not the case in SYCL device code.
+
+SYCL specialization constants (SYCL Section 4.9.5.) can be used to represent
+constants whose values can be set dynamically during the execution of a SYCL
+application, but that will not change when a SYCL kernel function is
+invoked. This way, specialization constants could be used to implement SYCL
+private arrays whose size is given during the execution of the SYCL
+application. There is no possible way of implementing this using `std::array`,
+as the size of such containers must be known at compile time, so we propose to
+define a new `private_alloca` function whose size is specified using SYCL
+specialization constants.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_SPEC_CONSTANT_LENGTH_ALLOCA` to one of the values defined
+in the table below. Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New Aspect for Specification Constant-Length Allocations
+
+This extension adds a new device aspect:
+
+[source,c++]
+----
+namespace sycl {
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_scla
+};
+} // namespace sycl
+----
+
+The `ext_oneapi_scla` aspect indicates that the device is capable of using the
+`private_alloca` API defined in the following sections.
+
+=== The SCLA API
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+template <typename ElementType, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+private_alloca(kernel_handler &kh);
+
+template <typename ElementType, std::size_t Alignment, auto &SpecName,
+          access::decorated DecorateAddress>
+private_ptr<ElementType, DecorateAddress>
+aligned_private_alloca(kernel_handler &kh);
+} // namespace sycl::ext::oneapi::experimental
+----
+
+_Mandates_: `ElementType` must be a cv-unqualified trivial type and `SpecName`
+must be a reference to a specialization constant of integral `value_type`. In
+the case of `aligned_private_alloca`, `Alignment` must be a power of 2
+fundamental alignment stricter than the alignment requirement of `ElementType`.
```

Comment:
I still can't parse this sentence.  The confusing part is "power of 2 fundamental alignment".  This makes it sound like the value must be the type's fundamental alignment raised to some power of 2.  Or, that the value must be 2^x where "x" is the type's fundamental alignment.  Neither of those things make any sense, though.

Reply:
I dropped the power of 2 claim. Alignments must be powers of 2 by definition, so the current statement works. If you like it, we can keep the current one or use the alternative equivalent:

> ... `Alignment` must be a fundamental alignment stricter than the alignment requirement of `ElementType`.

Both work for me

Reply:
I changed the condition to allow more values here. We no longer require fundamental alignments.

Reply:
> @v-klochkov does the current state look good to you?

The current design looks Ok/good to me.

Just some addition/idea. Originally I thought about something like this:
```C++
// Assuming 'DecoratedAddress' parameter may have default value.
// If DecorateAddress must be specified always, then the code is same, just remove '= yes' below.
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress = yes,
	  std::size_t Alignment = std::alignment_of_v<ElementType>>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, auto &SpecName,
          std::size_t Alignment = std::alignment_of_v<ElementType>,
	  access::decorated DecorateAddress = yes>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);
```
The reasoning:
`ElementType` and `SpecName`(aka Size) must always be specified and cannot be skipped.
The API variant shown above would be convenient if someone already has a call
```C++
auto P1 = private_alloca<float, size>(h);
auto P2 = private_alloca<float, size, access::decorated::no>(h);
```
and need to add alignment, then it is enough to simply it as 3rd or 4th template param without renaming:
```C++
// Now all pointers are 128-byte aligned.
auto P1 = private_alloca<float, size, 128>(h);
auto P2 = private_alloca<float, size, access::decorated::no, 128>(h);
auto P3 = private_alloca<float, size, 128, access::decorated::no>(h);
```

Reply:
Having a different name `aligned_private_alloca` has the advantage of making the aligned-ness extra highlighted. So, either way is good to me.

Reply:
I went with the `aligned_*` signature to follow SYCL and C++ existing library functions. I don't have a strong preference here either. @sommerlukas @Naghasan any preference?

Reply:
```c++
// Assuming 'DecoratedAddress' parameter may have default value.
// If DecorateAddress must be specified always, then the code is same, just remove '= yes' below.
template <typename ElementType, auto &SpecName,
          access::decorated DecorateAddress = yes,
	  std::size_t Alignment = std::alignment_of_v<ElementType>>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);

template <typename ElementType, auto &SpecName,
          std::size_t Alignment = std::alignment_of_v<ElementType>,
	  access::decorated DecorateAddress = yes>
private_ptr<ElementType, DecorateAddress>
private_alloca(kernel_handler &kh);
```

If I understand you @v-klochkov  correctly that you would propose to have both templates, this unfortunately wouldn't work due to ambiguous call in the case of `auto P1 = private_alloca<float, size>(h);` (https://godbolt.org/z/93dn19Gfj). 

I think we should go with the `aligned_*` signature as it follows C++ & SYCL and also makes the aligned-ness explicit.

Reply:
Okey, looking at your proposal a second time: having two signatures with the same name and your proposed usages would not work, as that would lead to ambiguous calls. As per the reasons I [exposed above](https://github.com/intel/llvm/pull/12660#issuecomment-1942112557), the two signatures approach might be more future proof. WDYT?